### ISSUE TITLE
C++

### ISSUE THREAD

**Issue by silent06** [XeCrippy/AchievementUnlockerV3#1]
Just wondering would you be willing to write this in C++? So people who don't have a PC can use an xex image?

**Comment by XeCrippy** [comment]
I might see if I can convert it when I have some time. It's not really something I have any interest in working on, I just put this out in case someone had any use for it. I'm sure there is an easier method I just haven't figured it out yet.

**Comment by XeCrippy** [comment]
So I tried to mess around with it a little bit and can't seem to figure out how to pattern search the xex for the function. I can find the pattern if I search a small range but I can't read through the whole xex to find it dynamically. 

From previous debugging it seems like you should be able to do it through xam but I could never get it to work. It uses MsgIORequest and MsgIORequestEx if that means anything. I'm not familiar enough with the c++ side really

**Comment by XeCrippy** [comment]
So I figured a little bit out but can't get XUserWriteAchievements working. I've done sonic adventure achievements in c++ but that's because i have the dev build and it uses another function that actually calls XUserWriteAchievements so it's different. Trying to figure out how to call it while calculating the address dynamically is giving me a headache. There are probably some xbox specific functions that would help but I'm lost. 

Here is how to find the function address anyways `void search_memory(void* start_address, size_t length, uint32_t target_value, uint32_t target_value_2) {
		uint8_t* mem = (uint8_t*)start_address;
		uint32_t test = 0x82000000;

		for (size_t i = 0; i < length - 3; i++) {
			if (*((uint32_t*)(mem + i)) == target_value) {
				if (i >= 8 && *((uint32_t*)(mem + i - 8)) == target_value_2) {
					std::ostringstream oss;
					oss << "Found XUserWriteAchievements at offset: 0x" << std::hex << test + i - 0x24;
					std::string result = oss.str();
					Utilities::Xam::XNotify(result);
					break;
				}
			}
		}
	}

	void achievementUnlocker::LocateXUserWriteAchievements() {
	     search_memory((void*)0x82000000, 0x1000000, 0x60840008, 0x38e00008);
	}`

**Comment by XeCrippy** [comment]
`void search_memory(void* start_address, size_t length, uint32_t target_value, uint32_t target_value_2) {
		uint8_t* mem = (uint8_t*)start_address;
		uint32_t test = 0x82000000;

		for (size_t i = 0; i < length - 3; i++) {
			if (*((uint32_t*)(mem + i)) == target_value) {
				if (i >= 8 && *((uint32_t*)(mem + i - 8)) == target_value_2) {
					std::ostringstream oss;
					oss << "Found XUserWriteAchievements at offset: 0x" << std::hex << test + i - 0x24;
					std::string result = oss.str();
					Utilities::Xam::XNotify(result);
					break;
				}
			}
		}
	}

	void achievementUnlocker::LocateXUserWriteAchievements() {
	     search_memory((void*)0x82000000, 0x1000000, 0x60840008, 0x38e00008);
	}`

**Comment by silent06** [comment]
xkelib has a function called "XUserWriteAchievements". Would that help? 

XBOXAPI
DWORD
WINAPI
XUserWriteAchievements(
    __in                            DWORD                       dwNumAchievements,
    __in_ecount( dwNumAchievements ) CONST XUSER_ACHIEVEMENT*   pAchievements,
    __inout_opt                     PXOVERLAPPED                pOverlapped
    );

**Comment by silent06** [comment]
xkelib has a function called "XUserWriteAchievements". Would that help? 

XBOXAPI
DWORD
WINAPI
XUserWriteAchievements(
    __in                            DWORD                       dwNumAchievements,
    __in_ecount( dwNumAchievements ) CONST XUSER_ACHIEVEMENT*   pAchievements,
    __inout_opt                     PXOVERLAPPED                pOverlapped
    );

**Comment by XeCrippy** [comment]
> xkelib has a function called "XUserWriteAchievements". Would that help? 
> 
> XBOXAPI
> DWORD
> WINAPI
> XUserWriteAchievements(
>     __in                            DWORD                       dwNumAchievements,
>     __in_ecount( dwNumAchievements ) CONST XUSER_ACHIEVEMENT*   pAchievements,
>     __inout_opt                     PXOVERLAPPED                pOverlapped
>     );
I've seen it but have no idea how to use it. With the c# version I just pointed pAchievements and pOverlapped to free memory in xam then the loop is to increase the value for pAchievments which I'm guessing is for the achievement IDs. So each time the loop runs it just tries the next value as the ID. This a bit beyond my scope.

**Comment by XeCrippy** [comment]
> xkelib has a function called "XUserWriteAchievements". Would that help?
> 
> XBOXAPI DWORD WINAPI XUserWriteAchievements( __in DWORD dwNumAchievements, __in_ecount( dwNumAchievements ) CONST XUSER_ACHIEVEMENT* pAchievements, __inout_opt PXOVERLAPPED pOverlapped );


Yeah so I'm an idiot lol. I figured it out now I'm pretty sure, just have to do a little more testing.

Thanks
